src/com/ms/silverking/cloud/dht/client/impl/ClientNamespace.java

```diff
@@ -33,326 +35,375 @@
-        return new AsynchronousNamespacePerspectiveImpl<K,V>(this, name,
-                                         new NamespacePerspectiveOptionsImpl<>(nspOptions, serializationRegistry));
+  }
+
+  @Override
+  public Namespace clone(String childName, long minVersion) throws NamespaceCreationException {
+    long creationTime;
+
+    creationTime = SystemTimeUtil.skSystemTimeSource.absTimeNanos();
+    switch (nsOptions.getVersionMode()) {
+    case CLIENT_SPECIFIED:
+      break;
+    case SEQUENTIAL:
+      throw new RuntimeException("Namespace.clone() not supported for version mode of SEQUENTIAL");
+    case SINGLE_VERSION:
+    case SYSTEM_TIME_MILLIS:
+    case SYSTEM_TIME_NANOS:
+      throw new RuntimeException(
+          "Namespace.clone() can only be supplied a version " + "for a version mode of CLIENT_SPECIFIED");
+    default:
+      throw new RuntimeException("Panic");
     }
+    return session.createNamespace(childName,
+        new NamespaceProperties(nsOptions, childName, this.name, minVersion, creationTime));
+  }
 
-	@Override
-	public <K, V> AsynchronousNamespacePerspective<K, V> openAsyncPerspective(
-			Class<K> keyClass, Class<V> valueClass) {
-        return new AsynchronousNamespacePerspectiveImpl<K,V>(this, name,
-                new NamespacePerspectiveOptionsImpl<>(getDefaultNSPOptions(keyClass, valueClass), serializationRegistry));
-	}
-	
-	@Override
-	public AsynchronousNamespacePerspective<String,byte[]> openAsyncPerspective() {
-		return openAsyncPerspective(DHTConstants.defaultKeyClass, DHTConstants.defaultValueClass);
-	}
-	
-	
-    @Override
-    public <K, V> SynchronousNamespacePerspective<K, V> openSyncPerspective(
-                                                                    NamespacePerspectiveOptions<K,V> nspOptions) {
-        return new SynchronousNamespacePerspectiveImpl<K, V>(this, name, 
-                new NamespacePerspectiveOptionsImpl<>(nspOptions, serializationRegistry));
+  public void linkTo(String parentName) throws NamespaceLinkException {
+    Namespace parent;
+
+    if (!nsOptions.getAllowLinks()) {
+      throw new NamespaceLinkException("Links not allowed for: " + name);
     }
 
-	@Override
-	public <K, V> SynchronousNamespacePerspective<K, V> openSyncPerspective(
-			Class<K> keyClass, Class<V> valueClass) {
-        return new SynchronousNamespacePerspectiveImpl<K, V>(this, name, 
-                new NamespacePerspectiveOptionsImpl<>(getDefaultNSPOptions(keyClass, valueClass), serializationRegistry));
-	}
-    
-	@Override
-	public SynchronousNamespacePerspective<String,byte[]> openSyncPerspective() {
-		return openSyncPerspective(DHTConstants.defaultKeyClass, DHTConstants.defaultValueClass);
-	}
-    
-    // misc.
-
-    @Override
-    public String toString() {
-        return name;
+    try {
+      parent = session.getNamespace(parentName);
+    } catch (NamespaceNotCreatedException nnce) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)